### ISSUE TITLE
save state / versioning

### ISSUE THREAD

**Issue by argyleink** [GoogleChromeLabs/ProjectVisBug#141]

**Comment by argyleink** [comment]
i pitched a visbug feature just like how Stylus works, really early on, and i still think it tells a great story and has a nice use case! i wanted to use "local overrides" in chrome, a way you can already persist changes and have them automatically applied to a page (no extension required!). essentially the story can be summed up as "make any webpage your own." 

have you seen this tool, it's a new and nice stylus https://www.styleurl.app. i have an opinion, that i'm waiting to be replaced with something better, which is that.. while this feature idea is awesome and tells a very compelling ux story, i gotta wonder why designers arent rockin stylus, recss, or styleurl today? why why why right? seems like tons of value, but apparently not? that logic path alone has halted me really attacking the "patch some css onto this page after load" story. have thoughts on this?

and yeah, overrides are different then versioning/saving state.. but they are related in many ways. perhaps i should break this out into 2 tickets?

**Comment by sw-yx** [comment]
hi Adam,

i am trying to make a "Save to Netlify" feature but am a bit lost about how to hack this together. i edit using visbug and the dom changes in the dev tools, but when i run `document.documentElement.outerHTML` or `document.body.innerHTML` i still see the old (original) DOM printed out. is there somewhere i can just "download" all the post edited HTML/CSS?

thanks
Shawn

**Comment by argyleink** [comment]
@sw-yx interesting! that's so sneaky/fishy sounding, i love it. i haven't tried what you're doing, lemme see if I can reproduce that. that's a cool feature idea though, treating visbug modified pages as an immutable state of the page (if i'm understanding your plan correctly)

**Comment by argyleink** [comment]
![screen shot 2018-12-16 at 3 45 21 pm](https://user-images.githubusercontent.com/1134620/50060519-fd92af80-0149-11e9-8503-9c798770dce3.png)

works for me (from dev console). where are you running `document.documentElement.outerHTML` from?

**Comment by sw-yx** [comment]
yes that is exactly the plan. save to an immutable deployment. its free on netlify so why not lol. part of me is also wary that youve probably got something similar cooking so i also wanted to post incase this was unnecessary work

i was running it in console. it definitely did not work, i tried quite a few different ways. im on vacation now but i’ll record a video sometime to demonstrate what i tried. anyway, if theres a button or method you wanna expose that just dumps all the html/css, then its a simple process of sending it to the [Netlify drop](https://app.netlify.com/drop) api for saving/sharable deployment and i figure thatd be great for your users

**Comment by argyleink** [comment]
yep, an idea like this would be great, like a netlify plugin that lets you sign in and use an auth key to push builds. i'm not working on anything like this 👍

i don't know of any easy ways to get work out, but there might be helpful stuff with the extension api's?   it's a tough one i think tho, dependencies need crawled and stuff? or urls need changed? i think the immutable visbug version deployment idea is a good one, but i dont have too many answers yet!

**Comment by sw-yx** [comment]
well there wouldnt even need to be authentication :) netlify is set up for anonymous deploys

and we dont need dependencies either if we’re just taking a literal snapshot of the visbugged DOM...

**Comment by Rafi993** [comment]
There might a way to solve this similar was draft-js has done rich text editing. We create a model (a JS object representation)  of the changes or just the current snapshot added to the DOM and store it so that it can be applied again.

**Comment by Rafi993** [comment]
Having a JS model will open up whole world of feature set like building additional plugins for this and undo redo downloading the changes and sharing them playing these changes as a timeline..... But also will make codebase easier to manage but it may require some work.

**Comment by sw-yx** [comment]
update - i'll own up that i just never got time to do this and am working on other projects now. i'll happily answer qtns for anyone interested in working on this.

**Comment by argyleink** [comment]
@Rafi993 i'd love to hear more about the strategy ya'll used, VisBug will need to eventually move to an abstraction of the nodes, where right now it works right against the DOM

i stumbled upon this tool recently and while it does similar things to Styleurl.app and others, I found it interesting that it uses your google account to stash the changes. I mean, most folks installing extensions probably are signed into Chrome. could be a clever way for VisBug to store state too, stash it onto a google account, and let it follow you that way. 

https://aminoeditor.com

**Comment by argyleink** [comment]
Still thinking about this, there's so many fun ways to share the work, it's hard to pick one:
1. diff to gist
1. dump current static state to netlify
1. export diff as zip
1. export page as site
1. your idea here

<br>

Here's the values I'm thinking should help guide us on our 1st solution:
1. doesnt require auth
1. can handle text edits / DOM changes
1. best if it can support toggling the style diff (think show/hide changes with 1 button)

your thoughts!?

**Comment by Rafi993** [comment]
Ramdajs does awesome work on this https://ramdajs.com/repl/ they encode everything in url. And can share the shortened url to others.

**Comment by Rafi993** [comment]
But still need to check if there any max length that url.

**Comment by CodyEakins** [comment]
@argyleink 
I like like idea of saving VisBug content changes to a gist, or as regular changes to a local git repository.

But how would you correctly save state when working with dynamic pages?

Take Youtube's homepage as an example. On their homepage, the featured videos are different for each user and they can change on each page request/refresh.

Obviously, if you tried to make changes to such a page, and compare the resulting HTML using Git, you'd get a bunch of false positives on which areas of the page have actually changed...

So, it seems to me that the only way to incorporate a "git diff" like feature would be to version each page request. Like, give the user an option to "lock" the base page before beginning design work with VisBug.

---

Side-note: I wonder if the native "Save Page As..." (or .mhtml) features of Chrome can be extended to help save pages and their state?

**Comment by argyleink** [comment]
> But how would you correctly save state when working with dynamic pages?

we could host the state of the static document when the save button is hit. the changes wont apply to the existing app's dynamic state, we'd be sharing the modified static document edited in a browser. at least, that's one way to "share or save your work." this method is specifically advantageous when showing DOM changes made, not just CSS changes.

> So, it seems to me that the only way to incorporate a git diff-like feature would be to version each page request.

i've got a branch where visbug can pass around an object between instances and re-apply them in order. this could let us store changes in the url or in a gist, and then reapply them to the page using a visbug plugin or command. in this case, it's less about a diff, and more about latest computed styles being reapplied to their target nodes. 

> side note
i think this very much could be used, but no one has prototyped it yet 😉 

hard to pick a strategy, they all have tradeoffs! 🙂 thanks for spinning up conversation about it 👍

**Comment by CodyEakins** [comment]
I think we're on the same page then. Especially when you mention:
> we'd be sharing the modified static document edited in a browser

This seems like the best way to save the VisBug "documents".

Making that static document, however, seems tricky... That's why I mentioned things like `Save Page As...` and `.mhtml` -- both are ways of copying a web page.

---

I'm interested in that branch you mentioned. Rather than store state exclusively with code (.html and css), did you have some sort of "action" system in mind? Like some kind of logger:
"VisBug added css rule to DOM node with ID: e9cb3ft"?

---

I want to be the one to help implement this feature, I just don't know where to start... Should I look into how to accurately save a page using Chrome? Or like make PR to add logging of VisBug actions to a file? Build something to compare VisBug static documents?

**Comment by Rafi993** [comment]
When we make changes using Visbug we can start recording changes in array of objects and when user opens up the web page again we apply those changes again to the page. For this to happen each operation that is possible using Visbug should probably be represented using action type and payload (like redux) with the payload being the location on which change was made and the change made (like text that was added).

**Comment by Rafi993** [comment]
So if user edits a text and then replaces image using visbug this local array would be

``` [{
  type: "EDIT_TEXT",
  location: "xyz",
  diff: "this is new text"
}, {
  type: "EDIT_IMAGE",
  location: "abc",
  diff: "imageURL"
}]
```

**Comment by CodyEakins** [comment]
@Rafi993 

I think having some sort of state store would be great while working on a site. Keep things fast and in memory.

However, when saving a file, I think the file itself is the stateful object. I think teams would rather pass around a HTML and CSS file rather than a JSON file... It depends.

Again, I don't think you can reliably recreate a document without versioning (making a copy of) the original page itself.

**Comment by Rafi993** [comment]
> @Rafi993 
> 
> I think having some sort of state store would be great while working on a site. Keep things fast and in memory.
> 
> However, when saving a file, I think the file itself is the stateful object. I think teams would rather pass around a HTML and CSS file rather than a JSON file... Well it depends, really, if you want multiple export formats?
> 
> Again, I don't think its possible to reliably recreate a document without versioning (making a copy of) the original page itself.

What I was essentially proposing was something similar to this https://youtu.be/feUYwoLhE_4

**Comment by Rafi993** [comment]
For storing state

**Comment by argyleink** [comment]
Here's a link to #371 which started work on a few things and is a good place to check what changed so all messages passed through a central location. by doing this we can get a visbug api to:
- publish node(s) selection
- publish node(s) is/was mutation payload
- set selection via payload
- mutate selection via payload

a few goals out of it:
- undo/redo via payloads
- allow external tools to leverage visbug mutation functionality
- allow external tools to "do something smart" with the diff/mutation
- allow a parent/child relationship to exist, where 1 central visbug could sync with others
- etc

it's pretty half baked at the moment, but does prove the concept. sounds like y'all have a good grip on the array of patterns to choose from to solve this, redux style payloads and the idea of time travel is along the lines of where the PR intend to head.  what kind of questions does all this raise for you?

**Comment by sw-yx** [comment]
i have left netlify but this may be a good one for @shortdiv or @tzmanics to pick up on in future if they so wish.

**Comment by CodyEakins** [comment]
@argyleink 

I'm game if you want to incorporate Redux or Flux-style architecture into this project.
It's definitely useful for keeping state organized, and it would make features like undo/redo possible...

---

Regarding state, however, I am more concerned about **creating "savable" projects**. 

If I am a designer, and I start tinkering with a page, I want to be able to come back the next day after closing my browser window. It has to be stable and portable.

This is why I think I'm going to look into things versioning pages, saving pages, and common file formats (export to .html/.css, .pdf, .doc, .json, etc.)

Hopefully I'm not alone or crazy in wanting this feature.

**Comment by CodyEakins** [comment]
@argyleink 

That sounds good. Makes sense to me.

The `.png` idea is interesting. In my use case, having the source code is more beneficial, but I could see some teams liking image exports better.

And I do want to attack this one. Let me know if you're available on a chat platform (such as Discord). I'd love to bounce some ideas off you in a more rapid-fire/spit-balling manner.

**Comment by argyleink** [comment]
fun fact, the `.png` is so large because it holds all the source code 😏 

best place to chat for me is [gitter](https://gitter.im/VisBug) or twitter, or start a PR? 

stoked you're attackin it! everything you need is sitting right there in the browser page right?! modifications are inline on that markup, excited to hash this out more with you. being able to share changes not as a screenshot would be really nice in many cases. thanks! 🙏

**Comment by CodyEakins** [comment]
Oh, really? I had no idea about that aspect of `.png` files.

I'll follow you on Gitter and Twitter then. Let's get after it!

**Comment by argyleink** [comment]
going to play around with this as a companion plugin https://github.com/gildas-lormeau/SingleFile
